### PR TITLE
Only use dtterm's extension where supported

### PR DESCRIPTION
**Pull request by jdebp**
`CSI` `8` `;` _h_ `;` _w_ `t` is an extension pioneered by dtterm, that not all terminals understand.  nvim is emitting it unconditionally, regardless of terminal type.  This is wrong, and has unfortunate side effects on some terminals.

When width and height are something such as (for example) 100 columns and 50 rows, this ends up emitting `CSI` `8` `;` `5` `0` `;` `1` `0` `0` `t`.  This can potentially be understood, by a terminal that does not understand the dtterm extension, as a regular DECSLPP sequence specifying a 100 _row_ screen size.  One ends up with an 100 by 100 screen, rather than the 100 by 50 one that was actually wanted.

So only emit the dtterm extension when nvim detects terminal types for terminals that are known to implement this extension, including `xterm*`, konsole, `teraterm*`, and `rxvt*`.

### CODE REVIEW

File: src/nvim/tui/tui.c [hunk 1]
```diff
@@ -974,14 +993,14 @@ static void fix_terminfo(TUIData *data)
     unibi_set_if_empty(ut, unibi_flash_screen, "\x1b[?5h$<20/>\x1b[?5l");
     unibi_set_if_empty(ut, unibi_enter_italics_mode, "\x1b[3m");
     unibi_set_if_empty(ut, unibi_to_status_line, "\x1b]2");
-  } else if (STARTS_WITH(term, "xterm")) {
+  } else if (data->term == kTermXTerm) {
```
Comment (brcolow) [inline src/nvim/tui/tui.c]: Nice reuse of the enum here

File: src/nvim/tui/tui.c [hunk 2]
```diff
@@ -974,14 +993,14 @@ static void fix_terminfo(TUIData *data)
...
     unibi_set_if_empty(ut, unibi_to_status_line, "\x1b]0;");
   } else if (STARTS_WITH(term, "screen") || STARTS_WITH(term, "tmux")) {
     unibi_set_if_empty(ut, unibi_to_status_line, "\x1b_");
     unibi_set_if_empty(ut, unibi_from_status_line, "\x1b\\");
   }
 
-  if (STARTS_WITH(term, "xterm") || data->term == kTermRxvt) {
+  if (data->term == kTermXTerm || data->term == kTermRxvt) {
```
Comment (brcolow) [inline src/nvim/tui/tui.c]: And here as well 👍

File: src/nvim/tui/tui.c [hunk 3]
```diff
@@ -423,9 +426,16 @@ static void tui_resize(UI *ui, int width, int height)
   ugrid_resize(&data->grid, width, height);
 
   if (!got_winch) {  // Try to resize the terminal window.
-    char r[16];  // enough for 9999x9999
-    snprintf(r, sizeof(r), "\x1b[8;%d;%dt", height, width);
-    out(ui, r, strlen(r));
+    // Only send this control sequence extension to terminal types that we know understand it.
```
Comment (brcolow) [inline src/nvim/tui/tui.c]: [The linter](https://travis-ci.org/neovim/neovim/jobs/228865508#L2992) complains about line length here and a couple other lines to line 434...if it is possible to shorten anything while retaining the meaning that would be swell - no big deal though.

File: src/nvim/tui/tui.c [hunk 4]
```diff
@@ -954,6 +964,15 @@ static TermType detect_term(const char *term, const char *colorterm)
...
   }
+  if (STARTS_WITH(term, "xterm")) {
+    return kTermXTerm;
+  }
+  if (STARTS_WITH(term, "dtterm")) {
+    return kTermDTTerm;
+  }
+  if (STARTS_WITH(term, "teraterm")) {
+    return kTermTeraTerm;
+  }
```
Comment (justinmk) [inline src/nvim/tui/tui.c]: Unless you have plans to use this granularity in a future patch, let's consider all of these to be xterm.

File: src/nvim/tui/tui.c [hunk 5]
```diff
@@ -974,14 +993,14 @@ static void fix_terminfo(TUIData *data)
...
+  } else if (data->term == kTermXTerm) {
     unibi_set_if_empty(ut, unibi_to_status_line, "\x1b]0;");
   } else if (STARTS_WITH(term, "screen") || STARTS_WITH(term, "tmux")) {
     unibi_set_if_empty(ut, unibi_to_status_line, "\x1b_");
     unibi_set_if_empty(ut, unibi_from_status_line, "\x1b\\");
   }
 
-  if (STARTS_WITH(term, "xterm") || data->term == kTermRxvt) {
+  if (data->term == kTermXTerm || data->term == kTermRxvt) {
     const char *normal = unibi_get_str(ut, unibi_cursor_normal);
```
Comment (justinmk) [inline src/nvim/tui/tui.c]: This won't get entered for kTermDTTerm/kTermTeraTerm. Is that intentional? What is the effect on those terminals?

**Comment by leonerd** [comment]
I would imagine also any terminals using `CSI t` in its more traditional capacity as `DECSLRM` would also get confused by this sequence.

File: src/nvim/tui/tui.c (hunk 5 above)
Comment (jdebp) [inline src/nvim/tui/tui.c]: Unchanged from what it was.  It wouldn't have been entered for `dtterm` or `teraterm*` terminal types before, either.

File: src/nvim/tui/tui.c (hunk 3 above)
Comment (jdebp) [inline src/nvim/tui/tui.c]: There you go.

**Comment by jdebp** [comment]
DECSLRM is `CSI` _l_ `;` _r_ `s ` .  That has its own overlaps, in particular with SCOSC; but they are not this one.

**Comment by jdebp** [comment]
An improved approach is to have an additional pretend `resize_screen` terminfo string capability that gets added into the real terminfo set for (only) these terminal types; rather than checking the terminal type again and again.

**Comment by justinmk** [comment]
@leonerd Are you in favor of this change? Or was there another approach you can suggest?

**Pull request closed without merging**